quickwit/quickwit-search/src/leaf_cache.rs

```diff
@@ -0,0 +1,288 @@
+// Copyright (C) 2023 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use quickwit_proto::{LeafSearchResponse, SearchRequest, SplitIdAndFooterOffsets};
+use quickwit_storage::MemorySizedCache;
+
+pub struct LeafSearchCache {
+    content: MemorySizedCache<CacheKey, LeafSearchResponse>,
+}
+
+impl LeafSearchCache {
+    pub fn new(capacity: usize) -> LeafSearchCache {
+        LeafSearchCache {
+            content: MemorySizedCache::with_capacity_in_bytes(
+                capacity,
+                &quickwit_storage::STORAGE_METRICS.partial_request_cache,
+            ),
+        }
+    }
+    pub fn get(
+        &self,
+        split_info: SplitIdAndFooterOffsets,
+        search_request: SearchRequest,
+    ) -> Option<LeafSearchResponse> {
+        let key = CacheKey::from_split_meta_and_request(split_info, search_request);
+        self.content.get(&key)
+    }
+
+    pub fn put(
+        &self,
+        split_info: SplitIdAndFooterOffsets,
+        search_request: SearchRequest,
+        result: LeafSearchResponse,
+    ) {
+        let key = CacheKey::from_split_meta_and_request(split_info, search_request);
+
+        self.content.put(key, result);
+    }
+}
+
+#[derive(Debug, Hash, PartialEq, Eq)]
+struct CacheKey {
+    split_id: String,
+    request: SearchRequest,
+    request_time_range: Range,
+}
+
+impl CacheKey {
+    fn from_split_meta_and_request(
+        split_info: SplitIdAndFooterOffsets,
+        mut search_request: SearchRequest,
+    ) -> Self {
+        let split_time_range = Range {
+            start: split_info.timestamp_start,
+            end: split_info.timestamp_end,
+        };
+        let request_time_range = Range {
+            start: search_request.start_timestamp,
+            end: search_request.end_timestamp,
+        }
+        .crop(&split_time_range);
+
+        search_request.start_timestamp = None;
+        search_request.end_timestamp = None;
+
+        CacheKey {
+            split_id: split_info.split_id,
+            request: search_request,
+            request_time_range,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+struct Range {
+    start: Option<i64>,
+    end: Option<i64>,
```

Comment:
It is crucial to comment what kind of interval this is supposed to be.
Is it closed or semi-open?
(The confusion is the source of many bugs in software)

Implementing https://doc.rust-lang.org/std/ops/trait.RangeBounds.html is nice too.